pkg/deviceshifu/deviceshifulwm2m/lwm2m/lwm2m.go

```diff
@@ -0,0 +1,387 @@
+		endpointName:    settings.EndpointName,
+		observeCallback: make(map[string]func(interface{})),
+		deviceTokenMap:  make(map[string]string),
+		settings:        settings,
+	}
+
+	router := mux.NewRouter()
+	if err := errors.Join(
+		router.Handle("/rd", mux.HandlerFunc(server.handleRegister)),
+		router.Handle("/rd/{deviceId}", mux.HandlerFunc(server.handleResourceUpdate)),
+	); err != nil {
+		return nil, err
+	}
+
+	router.DefaultHandle(mux.HandlerFunc(func(w mux.ResponseWriter, r *mux.Message) {
+		token := r.Token()
+		deviceId, exists := server.deviceTokenMap[string(token)]
+		if exists {
+			if fn, exists := server.observeCallback[deviceId]; exists {
+				data, err := io.ReadAll(r.Body())
+				if err != nil {
+					_ = w.SetResponse(codes.BadRequest, message.TextPlain, bytes.NewReader([]byte("failed to read body")))
+					return
+				}
+				fn(string(data))
+			}
+		}
+	}))
+
+	router.Use(loggingMiddleware)
+	server.router = router
+	return server, nil
+}
+
+func (s *Server) Run() error {
+	switch *s.settings.SecurityMode {
+	case v1alpha1.SecurityModeDTLS:
+		return s.startDTLSServer()
+	default:
+		logger.Infof("securityMode not set or not support, using none security mode")
+	// default using none security mode
+	case v1alpha1.SecurityModeNone:
+	}
+	return s.startUDPServer()
+}
+
+func (s *Server) startDTLSServer() error {
+	switch *s.settings.DTLSMode {
+	case v1alpha1.DTLSModePSK:
+		serverOptions := []dtlsServer.Option{
+			options.WithMux(s.router),
+			options.WithContext(context.Background()),
+			options.WithKeepAlive(10, time.Minute*10, func(cc *udpClient.Conn) {}),
+		}
+
+		server := dtlsServer.New(serverOptions...)
+
+		cipersuites, err := CipherSuiteStringsToCodes(s.settings.CipherSuites)
+		if err != nil {
+			return err
+		}
+
+		psk, err := hex.DecodeString(*s.settings.PSKKey)
+		if err != nil {
+			return err
+		}
+
+		dtlsConfig := dtls.Config{
+			PSK: func(hint []byte) ([]byte, error) {
+				return []byte(psk), nil
+			},
+			PSKIdentityHint: []byte(*s.settings.PSKIdentity),
+			CipherSuites:    cipersuites,
+		}
+
+		l, err := net.NewDTLSListener("udp4", ":5684", &dtlsConfig)
+		if err != nil {
+			return err
+		}
+
+		return server.Serve(l)
+	case v1alpha1.DTLSModeRPK:
+		fallthrough
+	case v1alpha1.DTLSModeX509:
+		return errors.New("not implemented")
+	default:
+		logger.Infof("dtlsMode not set, using none security mode")
+		// default using none security mode
+	}
+
+	return s.startUDPServer()
+}
+
+func (s *Server) startUDPServer() error {
+	serverOptions := []udpServer.Option{
+		options.WithMux(s.router),
+		options.WithContext(context.Background()),
+		options.WithKeepAlive(10, time.Minute*10, func(cc *udpClient.Conn) {
+			logger.Error("inactive connection")
+		}),
+	}
+
+	server := udpServer.New(serverOptions...)
+	conn, err := net.NewListenUDP("udp", ":5683")
+	if err != nil {
+		return err
+	}
+	return server.Serve(conn)
+}
+
+func (s *Server) handleRegister(w mux.ResponseWriter, r *mux.Message) {
```

Comment:
can we also parse objects and object instances from the register message?

Reply:
yes we can. But we read configmap to do it

Reply:
Let's add a todo here for that and create an issue